relay-server/src/services/buffer/envelope_stack/file_backed.rs

```diff
@@ -0,0 +1,625 @@
+
+/// The version of the envelope stack file format.
+const CURRENT_FILE_VERSION: u8 = 1;
+
+/// The size of the version field in bytes.
+const VERSION_FIELD_BYTES: u64 = 1;
+/// The size of the total count field in bytes.
+const TOTAL_COUNT_FIELD_BYTES: u64 = 4;
+/// The size of the envelope size field in bytes.
+const ENVELOPE_SIZE_FIELD_BYTES: u64 = 8;
+/// The size of the header for each envelope.
+const FILE_HEADER_SIZE_BYTES: u64 =
+    VERSION_FIELD_BYTES + TOTAL_COUNT_FIELD_BYTES + ENVELOPE_SIZE_FIELD_BYTES;
+
+/// An error returned when doing an operation on [`FileBackedEnvelopeStack`].
+#[derive(Debug, thiserror::Error)]
+pub enum FileBackedEnvelopeStackError {
+    #[error("failed to perform I/O operation: {0}")]
+    Io(#[from] io::Error),
+
+    #[error("failed to work with envelope: {0}")]
+    Envelope(#[from] EnvelopeError),
+
+    #[error("failed to get file from the store: {0}")]
+    EnvelopeStore(#[from] FileBackedEnvelopeStoreError),
+
+    #[error("the version of the entry in the file is not valid")]
+    InvalidVersion,
+
+    #[error("file corruption detected: {0}")]
+    Corruption(String),
+}
+
+/// An envelope stack that writes and reads envelopes to and from disk files.
+///
+/// Each [`FileBackedEnvelopeStack`] corresponds to a file on disk, named with the pattern
+/// `own_key-sampling_key`. The envelopes are appended to the file in a custom binary format.
+///
+/// The file format is as follows (from start to end):
+/// - `envelope_bytes` (variable length)
+/// - `version` (1 byte)
+/// - `total_count` (4 bytes, u32 in little-endian)
+/// - `envelope_size` (8 bytes, u64 in little-endian)
+///
+/// The rationale behind this layout is that we seek to the version and read to the right to get
+/// all the required metadata and not read more that it's needed. Then, once we know the size of
+/// the envelope, we seek back metadata size + envelope size.
+///
+/// This structure allows for efficient reading from the end of the file and
+/// updating of the total count when pushing or popping envelopes.
+#[derive(Debug)]
+pub struct FileBackedEnvelopeStack {
+    project_key_pair: ProjectKeyPair,
+    envelope_store: Arc<Mutex<FileBackedEnvelopeStore>>,
+}
+
+impl FileBackedEnvelopeStack {
+    /// Creates a new [`FileBackedEnvelopeStack`] instance.
+    pub fn new(
+        project_key_pair: ProjectKeyPair,
+        envelope_store: Arc<Mutex<FileBackedEnvelopeStore>>,
+    ) -> Self {
+        Self {
+            project_key_pair,
+            envelope_store,
+        }
+    }
+}
+
+impl EnvelopeStack for FileBackedEnvelopeStack {
+    type Error = FileBackedEnvelopeStackError;
+
+    async fn push(&mut self, envelope: Box<Envelope>) -> Result<(), Self::Error> {
+        let mut envelope_store = self.envelope_store.lock().await;
+        let file = envelope_store.get_file(self.project_key_pair).await?;
+
+        relay_statsd::metric!(timer(RelayTimers::BufferSpool), {
+            append_envelope(file, &envelope).await?;
```

Comment:
nit: consistent naming

```suggestion
            push_envelope(file, &envelope).await?;
```